src/wms/src/main/java/org/geoserver/wms/clip/ClippedFeatureSource.java

```diff
@@ -0,0 +1,100 @@
+import org.geotools.referencing.CRS;
+import org.geotools.util.logging.Logging;
+import org.locationtech.jts.geom.Geometry;
+import org.opengis.filter.Filter;
+import org.opengis.referencing.crs.CoordinateReferenceSystem;
+import org.opengis.referencing.operation.MathTransform;
+
+/** @author ImranR */
+public class ClippedFeatureSource extends DecoratingFeatureSource {
+    static final Logger LOGGER = Logging.getLogger(ClippedFeatureSource.class.getCanonicalName());
+    static final ClipProcess clipProcess = new ClipProcess();
+
+    Geometry clip;
+
+    public ClippedFeatureSource(FeatureSource delegate, Geometry clipGeometry) {
+        super(delegate);
+        this.clip = reproject(delegate.getInfo().getCRS(), clipGeometry);
+    }
+
+    private Geometry reproject(CoordinateReferenceSystem fsCRS, Geometry clipGeom) {
+        // re-project if required
+        try {
+            CoordinateReferenceSystem geomCRS = CRS.decode("EPSG:" + clipGeom.getSRID());
+            if (CRS.isTransformationRequired(geomCRS, fsCRS)) {
+                MathTransform mt = CRS.findMathTransform(geomCRS, fsCRS);
+                clipGeom = JTS.transform(clipGeom, mt);
+                clipGeom.setSRID(CRS.lookupEpsgCode(fsCRS, false));
+            }
+        } catch (Exception e) {
+            LOGGER.log(Level.SEVERE, "failed to reproject " + clipGeom.toText());
+            LOGGER.log(Level.SEVERE, e.getMessage(), e);
+        }
+
+        return clipGeom;
+    }
+
+    @Override
+    public ReferencedEnvelope getBounds() throws IOException {
+        return this.getFeatures().getBounds();
+    }
+
+    @Override
+    public ReferencedEnvelope getBounds(Query query) throws IOException {
+        return this.getFeatures(query).getBounds();
+    }
+
+    @Override
+    public int getCount(Query query) throws IOException {
+        return this.getFeatures(query).size();
+    }
+
+    @Override
+    public FeatureCollection getFeatures() throws IOException {
+        FeatureCollection fc = super.getFeatures();
+        if (SimpleFeatureCollection.class.isAssignableFrom(fc.getClass())) {
+            return getClippedCollection((SimpleFeatureCollection) fc, clip);
+        }
+        return fc;
+    }
+
+    @Override
+    public FeatureCollection getFeatures(Filter filter) throws IOException {
+        FeatureCollection fc = super.getFeatures(filter);
+        if (SimpleFeatureCollection.class.isAssignableFrom(fc.getClass())) {
+            return getClippedCollection((SimpleFeatureCollection) fc, clip);
+        }
+        return fc;
+    }
+
+    @Override
+    public FeatureCollection getFeatures(Query query) throws IOException {
+        FeatureCollection fc = super.getFeatures(query);
+        if (SimpleFeatureCollection.class.isAssignableFrom(fc.getClass())) {
+            return getClippedCollection((SimpleFeatureCollection) fc, clip);
+        }
+        return fc;
+    }
+
+    private static synchronized FeatureCollection getClippedCollection(

```

Comment:
Why static?

Reply:
@aaime To avoid instantiating a new ClipProcess for each OWS request separately.